pkg/frontend/querymiddleware/query_limiter_test.go

```diff
@@ -0,0 +1,376 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/user"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	promtest "github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/pkg/util/globalerror"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+func TestQueryLimiterMiddleware_RangeAndInstantQuery(t *testing.T) {
+	// All tests run with same query
+	query := "rate(metric_counter[5m])"
+
+	tests := []struct {
+		name                     string
+		limits                   mockLimits
+		expectSecondQueryBlocked bool
+	}{
+		{
+			name:   "empty limits should not block",
+			limits: mockLimits{},
+		},
+		{
+			name: "non-empty limit should block",
+			limits: mockLimits{limitedQueries: []*validation.LimitedQuery{
+				{Query: query, AllowedFrequency: time.Minute},
+			}},
+			expectSecondQueryBlocked: true,
+		},
+		{
+			name: "short max frequency should not block",
+			limits: mockLimits{limitedQueries: []*validation.LimitedQuery{
+				{Query: query, AllowedFrequency: time.Nanosecond},
```

Comment:
We can't actually use nanosecond TTLs in practice. Since we can advance "now" for the mock cache by any amount, might as well use a TTL that we could use in practice.